bin/cron/stop_inactive_adhoc_instances

```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/env ruby
+
+require_relative '../../dashboard/config/environment'
+require 'aws-sdk'
+require 'cdo/chat_client'
+
+StackStatus = Struct.new(
+  :name,
+  :cloudformation_status,
+  :creation_time,
+  :activity,
+  :current_instance_status,
+  :previous_instance_status,
+  :instance_status_change_reason
+)
+
+# set to :warn to notify that an adhoc environment should be stopped due to inactivity
+# set to :stop_instance to stop inactive instances
+EXECUTION_MODE = :warn
+
+# number of http requests received by environment within activity period
+# below which environment is considered to be inactive
+ACTIVITY_THRESHOLD = 1000
+
+# window of time within which activity is measured
+ACTIVITY_PERIOD_DAYS = 7
+
+# no activity is required within the last few days to protect new instances from being terminated
+GRACE_PERIOD_DAYS = 3
+
+# This script stops AWS EC2 instances being used to run adhoc environments if they have not been used recently:
+#
+# 1) for each AWS CloudFormation Stacks with 'adhoc' at the beginning of the name
+#   a) get AWS CloudWatch Metrics to determine if it has been inactive for more than a configurable time period
+#   b) stop instance if it activity is below threshold
+# 2) publish final status to slack cron-job (list of all adhoc environments and which ones were stopped)
+def main
+  ChatClient.message 'cron-daily', 'Beginning to stop inactive adhoc instances'
+  job_status = Array.new
+
+  cloudformation_client = Aws::CloudFormation::Client.new
+  cloudformation_resource = Aws::CloudFormation::Resource.new(client: cloudformation_client)
+  cloudwatch_client = Aws::CloudWatch::Client.new
+
+  cloudformation_resource.stacks.select {|stack| stack.name.starts_with?('adhoc-')}.each do |stack|
+    begin
+      status = StackStatus.new(stack.name, stack.stack_status, stack.creation_time)
+
+      # get Unicorn active HTTP request count CloudWatch metric for the current adhoc stack
+      metric = Aws::CloudWatch::Metric.new(namespace: 'Unicorn', name: 'active', client: cloudwatch_client)
+      unicorn_metric = metric.get_statistics(
+        {
+          dimensions: [
+            {
+              name: 'Environment',
+              value: 'adhoc',
+            },
+            {
+              name: 'Host',
+              value: "#{stack.name}.cdn-code.org",
+            }
+          ],
+          start_time: Time.now - (ACTIVITY_PERIOD_DAYS + GRACE_PERIOD_DAYS).days,
```

Comment:
Is this logic correct for applying a grace period? It seems to me that for a brand new instance, this would count the events from 10 days ago through 3 days ago, which will be 0, and the instance will be stopped. Maybe I'm misunderstanding?

Reply:
Good point.  This logic needs to be completely redesigned.

Reply:
@ewjordan I've updated and simplified the grace period logic.  Can you take another look?